teps/0042-taskrun-breakpoint-on-failure.md

```diff
@@ -0,0 +1,132 @@
+
+The scope of this TEP does not exceed any other kind of breakpointing or debugging apart from breakpointing on failure and failure itself.
+
+### Use Cases (optional)
+
+This enhancement can allow a more interactive way of working with TaskRuns. CLI or IDE extensions could allow users to run a 
+Task with breakpoint-on-failure and while the logs are being streamed, the user can be dropped to a shell in the breakpoint
+failed step and allow the user to debug as they develop.
+
+## Requirements
+
+* To enable breakpoint on failure of a step, it would be necessary to update the Entrypoint to support lifecycle
+changes when breakpoint on failure has been enabled.
+
+* The step environment would have to have scripts which the user can use to clear the breakpoint or
+mark the step as a success or a failure.
+
+* This would be paving the way integrating with IDE debuggers. 
+
+## Proposal
+
+TaskRuns will add a new spec called `breakpoint` under which `onFailure` can be set to true to enable breakpointing
+on failure.
+
+### Controlling step lifecycle
+
+#### Failure of a Step
+
+*What happens on failure of a step ? (Overview)*
+
+The entrypoint binary is used to manage the lifecycle of a step. Steps are aligned beforehand by the TaskRun controller
+allowing each step to run in a particular order. This is done using `-wait_file` and the `-post_file` flags. The former 
+let's the entrypoint binary know that it has to wait on creation of a particular file before starting execution of the step.
+And the latter provides information on the step number and signal the next step on completion of the step.
+
+On success of a step, the `-post-file` is written as is, signalling the next step which would have the same argument given
+for `-wait_file` to resume the entrypoint process and move ahead with the step. 
+
+On failure of a step, the `-post_file` is written with appending `.err` to it denoting that the previous step has failed with
+and error. The subsequent steps are skipped in this case as well, marking the TaskRun as a failure.
+
+#### Halting a Step on failure
+
+*Adding functionality to the entrypoint to handle step failures*
+
+The failed step writes `<step-no>.err` to `/tekton/tools` and stops running completely. To be able to debug a step we would
+need it to continue running (not exit), not skip the next steps and signal health of the step. By disabling step skipping, 
+stopping write of the `<step-no>.err` file and waiting on a signal by the user to disable the halt, we would be simulating a 
+"breakpoint".
+
+In this breakpoint, which is essentially a limbo state the TaskRun finds itself in, the user can interact with the step 
+environment using a CLI or an IDE. In this state the entrypoint is waiting on creation of the `<post-file>.breakpointexit` file which would mark the end of the debug session and the container would exit. eg: If step 0 is in debug more, we would have to write 0.breakpointexit to /tekton/tools to make the container exit.
```

Comment:
ah okay, so the entrypoint would be waiting (indefinitely) for a file to come into existence that would tell it to continue executing?

two questions:
a. after the breakpointexit file is created, with the entrypoint exit with an error (i.e. the exit code of the failed step?)
b. id like to understand a bit more about what would happen after this point - i really like that you've decoupled what the entrypoint binary does from whatever tooling built to interact with this might do, but i also feel like i need to understand a bit more of what that interaction might look like in order to understand if these entrypoint changes will be enough

i would expect as a user of this feature i would be able to:
1. set a breakpoint (right now, only on failure - some day, on a specific step, and other conditions like you mentioned!)
2. breakpoint which would be set via the taskrun, so the IDE or tool or whatever can take create of creating that
3. taskrun would execute as a pod, each container would run
4. some container fails - the entrypoint stops and waits
5. at this point, what would the user be doing? im thinking things like: looking at the state of the taskrun (e.g. what params it has, what results have been emitted, maybe logs), maybe connecting a shell and looking in the container (do we need to do something to support that or can we rely on k8s tools to take care of this?)

haha okay after writing this out i think my only question is really around (5) and if we need/want to do anything to support the "attaching a shell" part - maybe we don't! or maybe for not it is good enough to direct folks to `kubectl exec` for that?

Reply:
> ah okay, so the entrypoint would be waiting (indefinitely) for a file to come into existence that would tell it to continue executing?

yes !

> haha okay after writing this out i think my only question is really around (5) and if we need/want to do anything to support the "attaching a shell" part - maybe we don't! or maybe for not it is good enough to direct folks to kubectl exec for that?

The "attaching a shell" part should be taken care of by whatever client is being used for debugging. 
Just how in logging a taskrun with `tkn taskrun logs -f`, `tkn` gets the logs from each container and then shows it to us as one stream of logs. Similarly, the debug experience will be based on the implementation in the client and not in the controller, as the controller helps with managing the lifecycle of the taskrun and nothing more.

It is possible to get the shell with `kubectl exec` when debug is enabled. The plan is to replicate the same logic in the CLI for the first UI implementation.